posts/2023/08/03/ansible-data-manipulation-with-a-filter.md

```diff
@@ -0,0 +1,99 @@
+
+## Starting Point
+
+To begin, we need a dataset to work on. For this we used data from the automation controller API, workflows; it gives nested data on the nodes in each workflow to loop around. The variable file used in this case [can be found in the repo](https://github.com/sean-m-sullivan/ansible_data_manipulation_examples/blob/main/workflows.yml).
+
+The goal is to find what is being used in the automation controller looping over nested lists. While this is not a very practical example, it does give a starting point for creating a filter to manipulate any dataset.
+
+## Filter Basics
+
+The bones of this filter were taken from [ansible.netcommon.pop_ace](https://github.com/ansible-collections/ansible.netcommon/blob/main/plugins/filter/pop_ace.py). The start of every filter has some required options, such as FilterModule, and in addition AnsibleFilterError is good for troubleshooting.
+
+    from ansible.errors import AnsibleFilterError
+
+
+The class invocation sets the code as a filter, and invokes the function to use for the filter. This sets the filter called “used” in the playbook, and the function to invoke. Note that the function and the filter name do not need to match.
+
+    class FilterModule(object):
+        def filters(self):
+            return {"example_filter": self.workflow_manip}
+
+Then there is the documentation section: This can contain inputs, examples and other metadata. This is also how the ansible-docs are populated.
+
+            EXAMPLES = r"""
+            - name: Transform Data
+                ansible.builtin.set_fact:
+                data_out: "{{ workflow_job_templates | example_filter }}"
+            """
+
+For the most part this should be standard information. While the documentation field is not required for filters, it is best practice to include it. While not shown here, the [linked example](https://github.com/sean-m-sullivan/ansible_data_manipulation_examples/blob/main/filter_plugins/example_filter.py) also includes commented out expected output, which is great for going back and troubleshooting in the future.
+
+## Setting things up
+
+
+The first thing to do is set the filter arguments for data coming in. In our case the variable data_in, and that the input is of type dict. It is also best to set the return variable as empty here and any other defaults that need defined.
+
+        def example_filter(self, data_in: dict):
+            workflow_data = {}
+            workflow_data["workflows"] = []
+            workflow_data["job_templates"] = []
+            workflow_data["inventory_sources"] = []
+            workflow_data["approval_nodes"] = []
+
+The next step is to do the actual data manipulation.
+
+## In the thick of it
+
+This is where we get to what we actually want to do, take data from a source, loop through it, and extract the data needed. As the data is contained in nested lists, there is an inner and outer loop to go through.
+
+   	for workflow in data_in:
```

Comment:
Same as before. Looks like the code block here is from spaces. It works, but it's not consistent with what we have in the other submissions (backticks or ~~~), not sure if it's a problem, but highlighting just in case.